### PR TITLE
Change dtypes for toy models

### PR DESCRIPTION
**Pull request by Alex-Vasile**
Addresses a closely few issues:
1. Deepseek was using fp32 weights instead of fp16 weights.
2. Toy models were incorrectly using the same dtype for the norm weights as for the other layers. E.g. llama8b using fp32 for ffn_norm and attn_norm but fp16 for everything else, currently implementation was using fp16 for all.

It also expands the deepseek tests to:
1. Use KL-divergence for comparing logit distributions rather than a direct 1-to-1 comparisons of each logits.
2.  Paramaterize on dtype.

Finally, it splits out a helper method in `TorchGenerator` to seperately preprocess tokens as the same operation is copy and pasted across countless tests.

### CODE REVIEW

File: sharktank/tests/models/deepseek/test_sharded.py [hunk 1]
```diff
@@ -89,9 +89,24 @@ def testParallelToySizedModelEagerVsUnsharded(
         target_batch.prefill()
         target_logits = target_batch.prefill_logits
 
-        torch.testing.assert_close(
-            target_logits, reference_logits, atol=2e-4, rtol=2e-2
-        )
+        try:
+            torch.testing.assert_close(
+                target_logits, reference_logits, atol=2e-4, rtol=2e-2
```
[the remark this answers is not in the document]
Comment (sogartar) [reply sharktank/tests/models/deepseek/test_sharded.py:94]: It depends on what is the relative error when the absolute error is high.

**Review by sogartar — COMMENTED** [review]
We want to have tests for both f32 and for f16.
It is a toy model and should work with all dtypes.
Also being a toy model it should really not use these large dimensions like
```
rope_dimension_count = 64
vocabulary_size = 256
qk_nope_head_dim = 128
```

The toy model is supposed to show that the math is correct and not that we are not saturating or other numerical instabilities due to large model size.

File: sharktank/tests/models/deepseek/test_sharded.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sogartar) [reply sharktank/tests/models/deepseek/test_sharded.py]: That is bad.

[the remark this answers is not in the document]
Comment (Alex-Vasile) [reply sharktank/tests/models/deepseek/test_sharded.py]: It was caused by incorrectly set dtypes in the toy model creation.

File: sharktank/sharktank/models/llama/testing.py [hunk 2]
```diff
@@ -115,24 +123,27 @@ def make_attention_moe_block_random_theta(
...
-    dtype: Optional[torch.dtype] = None,
+    dtype_lo: torch.dtype = torch.float32,
+    dtype_hi: torch.dtype = torch.float32,
 ) -> Theta:
     if vocab_size is None:
         vocab_size = config.hp.vocab_size
-    if dtype is None:
-        dtype = config.dtype
+    if dtype_lo is None:
+        dtype_lo = config.dtype
```
Comment (sogartar) [inline sharktank/sharktank/models/llama/testing.py:132]: Should also set `dtype_hi` if not set.

**Review by sogartar — APPROVED** [review]
Thank you.
One remark I have is if it makes sense to rename the dtypes?
For example to `dtype` which is the main dtype of the model and `dtype_high_precision`/`dtype_high`. That is usually what we have. Some ops need to resort to a higher precision.
Of course with quantization things get a lot more complicated, but we are not there yet with test data generation.

File: sharktank/sharktank/layers/testing.py [hunk 3]
```diff
@@ -46,7 +47,7 @@ def make_llama_attention_block_theta(
             ),
             "attn_norm.weight": DefaultPrimitiveTensor(
                 name=f"blk.{block_idx}.attn_norm.weight",
-                data=make_rand_torch((embedding_length), dtype=dtype),
+                data=make_rand_torch((embedding_length), dtype=dtype_norm),
```
[the remark this answers is not in the document]
Comment (Alex-Vasile) [reply sharktank/sharktank/layers/testing.py:50]: Yeah, all of the RMS norm layers use fp32 in our 16bit quantization (bfloat16 native)

File: sharktank/sharktank/utils/load_llm.py [hunk 4]
```diff
@@ -43,13 +43,19 @@ def block_seq_stride(self) -> int:
     def preprocess_prompts(
```
Comment (archana-ramalingam) [inline sharktank/sharktank/utils/load_llm.py:43]: I understand the need to split this function. But now this essentially is just doing `self.tokenizer._encode` and printing log. Can't we add a fn attribute to enable/disable encoding?

File: sharktank/sharktank/layers/testing.py [hunk 5]
```diff
@@ -316,50 +319,52 @@ def make_random_ffn_theta(
...
     block_idx: int,
     in_dim: int,
     expert_hidden_dim: int,
     num_experts: int,
     with_ffn_norm: bool = True,
     num_shared_experts: int = 0,
     with_layer_output_norm: bool = False,
-    dtype: torch.dtype | None = None,
+    dtype_lo: torch.dtype,
+    dtype_hi: torch.dtype,
```
[the remark this answers is not in the document]
Comment (Alex-Vasile) [reply sharktank/sharktank/layers/testing.py]: Changed names to dtype_norm since that's all it's used for and dtype_rest.
